oak/server/rust/oak_runtime/src/node/grpc_server.rs

```diff
@@ -0,0 +1,275 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use log::{error, info, warn};
+
+use std::{
+    thread,
+    thread::{spawn, JoinHandle},
+};
+use futures_executor::block_on;
+use protobuf::Message;
+use protobuf::well_known_types::Any;
+
+use oak_abi::{ChannelReadStatus, OakStatus};
+use oak::grpc::GrpcRequest;
+
+use crate::Handle;
+use crate::{NodeId, RuntimeRef};
+
+pub struct GrpcServerNode {
+    config_name: String,
+    runtime: RuntimeRef,
+    node_id: NodeId,
+    writer: Handle,
+    thread_handle: Option<JoinHandle<()>>,
+}
+
+impl GrpcServerNode {
+    /// Creates a new [`GrpcServerNode`] instance, but does not start it.
+    pub fn new(
+        config_name: &str,
+        runtime: RuntimeRef,
+        node_id: NodeId,
+        writer: Handle,
+    ) -> Self {
+        Self {
+            config_name: config_name.to_string(),
+            runtime,
+            node_id,
+            writer,
+            thread_handle: None,
+        }
+    }
+}
+
+impl super::Node for GrpcServerNode {
+    fn start(&mut self) -> Result<(), OakStatus> {
+        let service = GrpcService {
+            config_name: self.config_name.clone(),
+            runtime: self.runtime.clone(),
+            node_id: self.node_id,
+            writer: self.writer,
+        };
+        // TODO(#770): Use `std::thread::Builder` and give a name to this thread.
+        let thread_handle = spawn(move || {
+            let pretty_name = format!("{}-{:?}:", service.config_name, thread::current());
+            // let address = format!("[::]:{}", port).parse()?;
+            let address = "127.0.0.1:8080".parse().unwrap();
+
+            let service_function = hyper::service::make_service_fn(move |_| {
+                let connection_service = service.clone();
+                async move {
+                    Ok::<_, hyper::Error>(hyper::service::service_fn(move |req| {
+                        let request_service = connection_service.clone();
+                        async move { request_service.serve(req).await }
+                    }))
+                }
+            });
+
+            let server = hyper::Server::bind(&address).serve(service_function);
+            if let Err(error) = block_on(server) {
+                error!("{} LOG: RPC server pseudo-node error: {}", pretty_name, error);
+            }
+
+            info!("{} LOG: exiting gRPC server thread", pretty_name);
+        });
+        self.thread_handle = Some(thread_handle);
+        Ok(())
+    }
+
+    fn stop(&mut self) {
+        if let Some(join_handle) = self.thread_handle.take() {
+            if let Err(err) = join_handle.join() {
+                error!("error while stopping gRPC server node: {:?}", err);
+            }
+        }
+    }
+}
+
+#[derive(Clone)]
+struct GrpcService {
+    config_name: String,
+    runtime: RuntimeRef,
+    node_id: NodeId,
+    writer: Handle,
+}
+
+impl GrpcService {
```

Comment:
Created node thread uses `Service` inside it for a Hyper server.
And `GrpcServerNode` contains a thread reference.
So it would require creating a custom (non-`derive`) `clone` function that doesn't copy this thread reference and sets it to `None`.
Not sure if it will be readable.

Reply:
Makes sense, but perhaps at least create the `GrpcService` instance as part of `GrpcServerNode` itself, rather than repeat all the fields?

Also I am confused by the naming, why is one called `service` and the other `server`? I think we should stick to one of them, and I think `server` would be appropriate.

Reply:
Done